### ISSUE TITLE
Bug: query inconsistent result between the old and new engine

### ISSUE THREAD

**Issue by yeya24** [thanos-community/promql-engine#200]
Input:

```
	load := `load 30s
http_requests_total{pod="nginx-1", series="1"} 1+1.1x40
http_requests_total{pod="nginx-2", series="2"} 2+2.3x50
http_requests_total{pod="nginx-3", series="3"} 6+0.8x60
http_requests_total{pod="nginx-4", series="3"} 5+2.4x50
http_requests_total{pod="nginx-5", series="1"} 8.4+2.3x50
http_requests_total{pod="nginx-6", series="2"} 2.3+2.3x50
`
```

Did an instant query using the query below and it caused different results for the two engines. The query is generated from my query fuzzer so probably it doesn't make sense. But it is interesting to see why two engines gave different results.

```
vector(-(-scalar(stdvar({series="3"} offset 26s)) > bool 0.8537454363931435) < bool time())
    testutil.go:91: engine_test.go:86: ""
        
        	exp: &promql.Result{Err:error(nil), Value:promql.Vector{promql.Sample{Point:promql.Point{T:1677955791693, V:1, H:(*histogram.FloatHistogram)(nil)}, Metric:labels.Labels{}}}, Warnings:storage.Warnings(nil)}
        
        	got: &promql.Result{Err:error(nil), Value:promql.Vector{}, Warnings:storage.Warnings(nil)}
        
        Diff:
        --- Expected
        +++ Actual
        @@ -1,2 +1,2 @@
        -(*promql.Result)({} => 1 @[1677955791693])
        +(*promql.Result)()
```

**Comment by mhoffm-aiven** [comment]
I'm interested: can you share your query fuzzer?

**Comment by yeya24** [comment]
```
scalar(avg_over_time({__name__="http_requests_total"}[3m])) > bool 0.9464749352949011
    testutil.go:91: engine_test.go:86: ""
        
        	exp: &promql.Result{Err:error(nil), Value:promql.Scalar{T:1678043228609, V:0}, Warnings:storage.Warnings(nil)}
        
        	got: &promql.Result{Err:error(nil), Value:promql.Scalar{T:1678043228609, V:NaN}, Warnings:storage.Warnings(nil)}
        
        Diff:
        --- Expected
        +++ Actual
        @@ -1,2 +1,2 @@
        -(*promql.Result)(scalar: 0 @[1678043228609])
        +(*promql.Result)(scalar: NaN @[1678043228609])
```

The old engine returns 0 but the new engine returns value NaN

**Comment by yeya24** [comment]
Another query.

```
ln(
  vector(
      (
          (time() > bool (scalar(clamp_max({series="1"}, 0.34725748033422527)) != bool 0.026108016140569366))
        != bool
          (0.1705142984178662 < bool scalar({pod="nginx-5"}))
      )
    >= bool
      0.49604491528174655
  )
)

    testutil.go:91: engine_test.go:86: ""
        
        	exp: &promql.Result{Err:error(nil), Value:promql.Vector{promql.Sample{Point:promql.Point{T:1678043387067, V:0, H:(*histogram.FloatHistogram)(nil)}, Metric:labels.Labels{}}}, Warnings:storage.Warnings(nil)}
        
        	got: &promql.Result{Err:error(nil), Value:promql.Vector{promql.Sample{Point:promql.Point{T:1678043387067, V:-Inf, H:(*histogram.FloatHistogram)(nil)}, Metric:labels.Labels{}}}, Warnings:storage.Warnings(nil)}
        
        Diff:
        --- Expected
        +++ Actual
        @@ -1,2 +1,2 @@
        -(*promql.Result)({} => 0 @[1678043387067])
        +(*promql.Result)({} => -Inf @[1678043387067])
```

This one is more interesting because the new engine returns `-Inf`.

**Comment by yeya24** [comment]
This seems a separate issue we have.

```
timestamp(vector(pi()))
    testutil.go:91: engine_test.go:86: ""
        
        	exp: &promql.Result{Err:error(nil), Value:promql.Vector{promql.Sample{Point:promql.Point{T:1678043516356, V:1.678043516356e+09, H:(*histogram.FloatHistogram)(nil)}, Metric:labels.Labels{}}}, Warnings:storage.Warnings(nil)}
        
        	got: &promql.Result{Err:error(nil), Value:promql.Vector{promql.Sample{Point:promql.Point{T:1678043516356, V:0, H:(*histogram.FloatHistogram)(nil)}, Metric:labels.Labels{}}}, Warnings:storage.Warnings(nil)}
        
        Diff:
        --- Expected
        +++ Actual
        @@ -1,2 +1,2 @@
        -(*promql.Result)({} => 1678043516.356 @[1678043516356])
        +(*promql.Result)({} => 0 @[1678043516356])
```

**Comment by yeya24** [comment]
@mhoffm-aiven Thanks for the interest. I will see if I can improve it a little bit more and check if I can open source it.

**Comment by MichaHoffmann** [comment]
> This seems a separate issue we have.
> 
> ```
> timestamp(vector(pi()))
>     testutil.go:91: engine_test.go:86: ""
>         
>         	exp: &promql.Result{Err:error(nil), Value:promql.Vector{promql.Sample{Point:promql.Point{T:1678043516356, V:1.678043516356e+09, H:(*histogram.FloatHistogram)(nil)}, Metric:labels.Labels{}}}, Warnings:storage.Warnings(nil)}
>         
>         	got: &promql.Result{Err:error(nil), Value:promql.Vector{promql.Sample{Point:promql.Point{T:1678043516356, V:0, H:(*histogram.FloatHistogram)(nil)}, Metric:labels.Labels{}}}, Warnings:storage.Warnings(nil)}
>         
>         Diff:
>         --- Expected
>         +++ Actual
>         @@ -1,2 +1,2 @@
>         -(*promql.Result)({} => 1678043516.356 @[1678043516356])
>         +(*promql.Result)({} => 0 @[1678043516356])
> ```

the timestamp function is just broken currently, i raised #204  for a discussion.

**Comment by MichaHoffmann** [comment]
I think whats broken is the "bool" modifier; that seems to be part of most examples and also kinda explains the `ln` one as that seems like `ln(1) vs ln(0)`

**Comment by yeya24** [comment]
```
		{
			name: "his",
			load: `load 30s
http_requests_total{pod="nginx-1", series="1"} 1+1.1x40
http_requests_total{pod="nginx-2", series="2"} 2+2.3x50
http_requests_total{pod="nginx-3", series="3"} 6+0.8x60
http_requests_total{pod="nginx-4", series="3"} 5+2.4x50
http_requests_total{pod="nginx-5", series="1"} 8.4+2.3x50
http_requests_total{pod="nginx-6", series="2"} 2.3+2.3x50
`,
			query: `(count by (pod, series) (vector(scalar({pod="nginx-1"} offset 1m55s))))`,
		},

=== RUN   TestInstantQuery/disableOptimizers=true/his/disableFallback=true
    testutil.go:91: engine_test.go:2616: ""
        
        	exp: &promql.Result{Err:error(nil), Value:promql.Vector{promql.Sample{Point:promql.Point{T:50000, V:1, H:(*histogram.FloatHistogram)(nil)}, Metric:labels.Labels{}}}, Warnings:storage.Warnings(nil)}
        
        	got: &promql.Result{Err:error(nil), Value:promql.Vector{}, Warnings:storage.Warnings(nil)}
        
        Diff:
        --- Expected
        +++ Actual
        @@ -1,2 +1,2 @@
        -(*promql.Result)({} => 1 @[50000])
        +(*promql.Result)()
```

`(count by (pod, series) (vector(scalar({pod="nginx-1"} offset 1m55s))))`
Another test case that I am able to see inconsistencies. Seems the `offset 1m55s` caused the issue.

**Comment by yeya24** [comment]
Will be fixed by #237
